tests/unit/modules/test_role.py

```diff
@@ -0,0 +1,141 @@
+
+class TestRole(ModuleTestCase):
+    def test_minimal_role_parameters(self, mocker):
+        sync_mock = mocker.patch.object(utils, 'sync')
+        sync_mock.return_value = True, {}
+        set_module_args(
+            name='test_role',
+            rules=[
+                dict(
+                    verbs=[],
+                    resources=[]
+                ),
+            ]
+        )
+
+        with pytest.raises(AnsibleExitJson):
+            role.main()
+
+        state, _client, path, payload, check_mode = sync_mock.call_args[0]
+        assert state == 'present'
+        assert path == '/roles/test_role'
+        assert payload == dict(
+            rules=[
+                dict(
+                    verbs=[],
+                    resources=[],
+                    resource_names=None,
+                )
+            ],
+            metadata=dict(
+                name='test_role',
+                namespace='default',
+            ),
+        )
+        assert check_mode is False
+
+    def test_all_role_parameters(self, mocker):
+        sync_mock = mocker.patch.object(utils, 'sync')
+        sync_mock.return_value = True, {}
+        set_module_args(
+            name='test_role',
+            state='present',
+            rules=[
+                dict(
+                    verbs=['get', 'list', 'create'],
+                    resources=['assets', 'entities'],
+                ),
+                dict(
+                    verbs=['list'],
+                    resources=['check'],
+                    resource_names=['my-check'],
+                )
+            ],
+        )
+
+        with pytest.raises(AnsibleExitJson):
+            role.main()
+
+        state, _client, path, payload, check_mode = sync_mock.call_args[0]
+        assert state == 'present'
+        assert path == '/roles/test_role'
+        assert payload == dict(
+            metadata=dict(
+                name='test_role',
+                namespace='default',
+            ),
+            rules=[
+                dict(
+                    verbs=['get', 'list', 'create'],
+                    resources=['assets', 'entities'],
+                    resource_names=None,
+                ),
+                dict(
+                    verbs=['list'],
+                    resources=['check'],
+                    resource_names=['my-check'],
+                )
+            ],
+        )
+
+    def test_failure(self, mocker):
+        sync_mock = mocker.patch.object(utils, 'sync')
+        sync_mock.side_effect = errors.Error("Bad error")
+        set_module_args(
+            name='test_role',
+            rules=[
+                dict(
+                    verbs=[],
+                    resources=[],
+                )
+            ],
+        )
+        with pytest.raises(AnsibleFailJson):
+            role.main()
+
+    def test_failure_invalid_verb(self, mocker):
+        sync_mock = mocker.patch.object(utils, 'sync')
+        sync_mock.return_value = True, {}
```

Comment:
Do we need those two lines? I would expect that parameter validation kicks in before we get busy interacting with the API.

Reply:
I was so happy that I found copy-paste remnant that I completely forgot to think about anything else ;) And now that I know why those two lines are here, I would like to suggest a minor adaptation. Instead of setting a return value in `sync` mock object, can we instead specify a side effect and raise an exception? For example

    sync_mock.side_effect = Exception("Validation should fail before we get to sync")

The main reason for my plead is: I am getting old and I will forget this discussion in abour an hour or so. And when the test fails in the future, I am sure the error message in that exception will be the most valuable thing ever ;)